### ISSUE TITLE
Use persistent child R sessions to process parse and diagnostics tasks

### ISSUE THREAD

**Issue by renkun-ken** [REditorSupport/languageserver#207]
Currently, we are using `callr::r_bg()` to create background R sessions to process tasks such as `parse_task`, `diagnostics_task`, and `resolve_task`. However, the overhead of launching new process is much more than parsing the document.

Unlike `resolve_task` which requires a fresh R session to get clean search paths to resolve packages, `parse_task` and `diagnostics_task` do not need fresh R sessions to process each task. The overhead could be starting an R session and loading `languageserver` and its dependencies.

I'm wondering if we could use persistent child R sessions to process parse and diagnostics tasks. See https://github.com/r-lib/callr#persistent-r-sessions.

Following are some benchmarks:

`script.R` is a script of 200 lines.

```r
parse_test <- function() {
  file <- "script.R"
  lines <- readLines(file)
  languageserver:::parse_document(file, lines)
}

r_bg_parse <- function() {
  r <- callr::r_bg(parse_test)
  r$wait()
  r$get_result()
}

persistent_r_parse <- function() {
  # R$call(parse_test)
  # while (TRUE) {
  #   if (!is.null(R$read())) {
  #     break
  #   }
  # }
  R$run(parse_test)
}

R <- callr::r_session$new()

microbenchmark::microbenchmark(
  parse_test(),
  r_bg_parse(),
  persistent_r_parse(),
  times = 100
)
```

```
Unit: milliseconds
                 expr       min        lq      mean    median        uq        max neval
         parse_test()  19.17827  20.23432  21.94873  20.88153  21.88695   35.96008   100
         r_bg_parse() 496.95088 545.88191 735.77912 576.86343 887.01842 4612.39674   100
 persistent_r_parse() 105.28276 111.75743 115.59718 115.08366 117.73012  198.30838   100
```

The approach we are using is `r_bg_parse()`. Its performance is quite bad and unstable compared to direct `parse_test()` and `persistent_r_parse()`.

**Comment by randy3k** [comment]
The task manager was originally written for lintr/diagnostic which could be quite slow. I actually am not sure if we need to run parse_document in `callr`. If parsing is quick (even for large files), we could put it in the main process.

For diagnostic tasks, it does make sense to have a persistent session per document/workspace.

**Comment by renkun-ken** [comment]
Actually, yesterday I implemented a version that put `parse_document` in main process, and for larger document (500 - 1000 lines of code), the delay is quite notable such as there's a delay of completion, notable delay in completion resolve, etc.

It looks like such parsing running in the background is almost a must especially for those who edit large documents, or otherwise the langaugeserver is almost unuseable.

**Comment by randy3k** [comment]
We might need to parse a large number of files for R package projects when the server starts, so we might need to be a bit flexible.

**Comment by renkun-ken** [comment]
I did some experiment where persistent R sessions are used to handle tasks and it does not seem to work as well as `r_bg` since current task manager has a unique advantage that the task for a document could be safely dropped by killing the task process to make way for latest task.

It is not easy to safely and stably interrupt the expression a background persistent R session is running as I tried although it can be done with `session$interrupt()` or `session$signal(2)`. But it will occasionally cause the session to hang or behave strangely yet the cause is unknown.

I guess we could only rely on the current implement at the moment which works well and significantly improves the availability of the languageserver.
